File: 611.ValidTriangleNumber/Solution.go
import "sort"

func triangleNumber(nums []int) int {
	if nums == nil || len(nums) == 0 {
		return 0
	}

	counter := 0
	//sort slice
	sort.Ints(nums)

	//iterate index of the 3rd edge, nums[3rd_idx] has to be larger than nums[2nd_idx] + nums[1st_idx]
	for third_edge_idx := 2; third_edge_idx < len(nums); third_edge_idx++ {
		first_edge_idx := 0
		second_edge_idx := third_edge_idx - 1
		//second_edge_idx, the right pointer while the first_edge_idx works as the left pointer
		//iterate over condition that two pointers do not overlap
		for first_edge_idx < second_edge_idx {
			//model the condition
			if nums[first_edge_idx]+nums[second_edge_idx] > nums[third_edge_idx] {
				counter += second_edge_idx - first_edge_idx
				second_edge_idx--
			} else {
				first_edge_idx++
			}
		}

	}
	return counter
}

File: 611.ValidTriangleNumber/Solution.java
import java.util.Arrays;

class Solution {
    public int triangleNumber(int[] nums) {
        if(nums == null || nums.length == 0)return 0;
        
        int counter = 0;
        //sort list
        Arrays.sort(nums);
        //iterate the array from the 2nd value
        
        for (int third_edge_idx=2; third_edge_idx<nums.length; third_edge_idx++){
            int first_edge_idx = 0;
            int second_edge_idx = third_edge_idx-1;
            while (first_edge_idx<second_edge_idx){
                if (nums[first_edge_idx] + nums[second_edge_idx] > nums[third_edge_idx]){
                    counter += second_edge_idx-first_edge_idx;
                    second_edge_idx--;
                }else{
                    first_edge_idx++;
                }
            }
            
        }
        return counter;  
    }
}